File: public/dsa/searching/square_root/storing_previous_result/square_root.py
from math import ceil


class Solution:
    def searchRoot(self, num: int) -> int:
        if num <= 1:
            return num
        l = 0
        r = num // 2
        while l < r:
            m = ceil((l + r) / 2)
            sq = m * m
            if sq > num:
                r = m - 1
            else:
                l = m
        return l


# tests-start
sol = Solution()
print(sol.searchRoot(16))
print(sol.searchRoot(12))
# tests-end

'''output
4
3
'''

File: public/dsa/searching/square_root/storing_previous_result/square_root.cpp
#include <bits/stdc++.h>
using namespace std;

int searchRoot(int num) {
    if (num == 0 || num == 1) {
        return num;
    }
    int low = 1, high = num / 2, ans = 1;
    while (low <= high) {
        long long mid = (low + high) / 2;
        long long sq = mid * mid;
        if (sq <= num) {
            ans = mid;
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }
    return ans;
}

// tests-start
int main() {
    cout << searchRoot(16) << endl;
    cout << searchRoot(12) << endl;
}
// tests-end

/*output
4
3
*/